cmd/internal/cli/build_linux.go

```diff
@@ -468,3 +477,146 @@ func getEncryptionMaterial(cmd *cobra.Command) (*cryptkey.KeyInfo, error) {
 
 	return nil, nil
 }
+
+func processDefs(args []string, argFile string, defs []types.Definition) ([]types.Definition, error) {
+	// --build-arg and content in --build-arg-file are applied to all stages
+	buildArgMap, err := readBuildArgs(args, argFile)
+	if err != nil {
+		return defs, err
+	}
+
+	// start replacing the variable defined in the definition file
+	for idx, def := range defs {
+		d, err := updateDef(&def, buildArgMap)
+		if err != nil {
+			return defs, fmt.Errorf("while updating the definition file with build args, err: %w", err)
+		}
+		defs[idx] = *d
+	}
+
+	types.UpdateDefinitionRaw(defs)
+
+	return defs, nil
+}
+
+func readBuildArgs(args []string, argFile string) (map[string]string, error) {
+	buildVarsMap := make(map[string]string)
+	if argFile != "" {
+		file, err := os.Open(argFile)
+		if err != nil {
+			return buildVarsMap, fmt.Errorf("while opening the file %s, err: %w", argFile, err)
+		}
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			text := scanner.Text()
+			k, v, err := getKeyVal(text)
+			if err != nil {
+				sylog.Warningf("Skipping the line, err: %v", err)
+				continue
+			}
+
+			buildVarsMap[k] = v
+		}
+
+		if err := scanner.Err(); err != nil {
+			return buildVarsMap, fmt.Errorf("while scanning the content of target file %s, err: %w", argFile, err)
+		}
+	}
+
+	for _, arg := range args {
+		k, v, err := getKeyVal(arg)
+		if err != nil {
+			sylog.Warningf("Skipping the line, err: %v", err)
+			continue
+		}
+
+		buildVarsMap[k] = v
+	}
+
+	return buildVarsMap, nil
+}
+
+func getKeyVal(text string) (string, string, error) {
+	if !strings.Contains(text, "=") {
+		return "", "", fmt.Errorf("text: %s is not `key=value` pair format", text)
+	}
+
+	matches := strings.SplitN(text, "=", -1)
+	if len(matches) != 2 {
+		return "", "", fmt.Errorf("text: %s is not `key=value` pair format", text)
+	}
+
+	key := strings.TrimSpace(matches[0])
+	if key == "" {
+		return "", "", fmt.Errorf("key field is missing in text: %s", text)
+	}
+	val := strings.TrimSpace(matches[1])
+	if val == "" {
+		return "", "", fmt.Errorf("value field is missing in text: %s", text)
+	}
+	return key, val, nil
+}
+
+var errNoChange = errors.New("text no change")
```

Comment:
How about "no change to text" instead.

Reply:
thanks! resolved in the latest commit.
https://github.com/apptainer/apptainer/pull/1344/commits/1842bc374a7db1c7a72fb8fa11d811d7de4e2ff4